### DISCUSSION TITLE
Can I specify language/JDK level?

### DISCUSSION THREAD

**Discussion by omega** [JetBrains/Qodana#68] [Q&A] [answered]
We run our apps with java 17, and use some java 17 features, like `Stream.toList()`, but Qodana seems to check with another java version?

What is even stranger is that the next line is a suggestion to replace `collect(toList())` with `toList()`, but the second is not from the sanity-checker perhaps?

<img width="1403" alt="image" src="https://user-images.githubusercontent.com/23787/152010141-6d29f875-71f4-45ec-9a03-d8f3c4bdf9eb.png">

**Comment by brichbash** [comment, marked as answer]
~~The information about the required jdk and language level is stored in `.idea/misc.xml` that is used and generated (if it is absent) in the project root by Qodana. You can create it before a step with Qodana. `misc.xml` should contain:~~
```
<?xml version="1.0" encoding="UTF-8"?>
<project version="4">
  <component name="ExternalStorageConfigurationManager" enabled="true" />
  <component name="FrameworkDetectionExcludesConfiguration">
    <file type="web" url="file://$PROJECT_DIR$" />
  </component>
  <component name="ProjectRootManager" version="2" languageLevel="JDK_17" project-jdk-name="openjdk-17" project-jdk-type="JavaSDK">
    <output url="file://$PROJECT_DIR$/out" />
  </component>
</project>
```
You can use other JDKs from the list in [the attached file](https://github.com/JetBrains/Qodana/files/7985734/jdks.json.zip) using names from `suggested_sdk_name` field.
**UPDATE:**
Since 2022.2 you can use `projectJDK` [property](https://www.jetbrains.com/help/qodana/configure-jdk.html#Configure+Qodana) in `qodana.yaml`.

**Reply by ropsahl** [reply]
I have tried this with jetbrains/qodana-jvm:2022.1-eap but I get:
WARN - #c.i.o.p.i.UnknownSdkInspectionCommandLineConfigurator - Failed to resolve JDK "openjdk-17": JDK "openjdk-17" is missing
And my .idea/misc.xml is then overwritten by default value.

**Reply by brichbash** [reply]
Could you please try using "17" instead of "openjdk-17"?

**Reply by ropsahl** [reply]
Yes, Thank you, that worked!
